src/wpilib/joystick.rs

```diff
@@ -0,0 +1,223 @@
+        match side {
+            JoystickSide::LeftHand => self.left_rumble = (value * RUMBLE_BASE as f32) as i32,
+            JoystickSide::RightHand => self.right_rumble = (value * RUMBLE_BASE as f32) as i32,
+        }
+        unsafe { HAL_SetJoystickOutputs(self.port as i32, self.outputs, self.left_rumble, self.right_rumble) };
+    }
+}
+
+pub struct XBoxController {
+    port: usize,
+    ds: &'static mut DriverStation,
+    outputs: i64,
+    left_rumble: i32,
+    right_rumble: i32,
+}
+
+impl XBoxController {
+    pub fn new(p: usize) -> XBoxController {
+        XBoxController { port: p, ds: DriverStation::instance(), outputs: 0i64, left_rumble: 0i32, right_rumble: 0i32 }
+    }
+
+    //buttons
+    pub fn get_a_button(&mut self) -> bool {
+        self.get_raw_button(1)
+    }
+
+    pub fn get_b_button(&mut self) -> bool {
+        self.get_raw_button(2)
+    }
+
+    pub fn get_x_button(&mut self) -> bool {
+        self.get_raw_button(3)
+    }
+
+    pub fn get_y_button(&mut self) -> bool {
+        self.get_raw_button(4)
+    }
+
+    pub fn get_bumper(&mut self, side: JoystickSide) -> bool {
+        match side {
+            JoystickSide::LeftHand => self.get_raw_button(5),
+            JoystickSide::RightHand => self.get_raw_button(6),
+        }
+    }
+
+    pub fn get_back_button(&mut self) -> bool {
+        self.get_raw_button(7)
+    }
+
+    pub fn get_start_button(&mut self) -> bool {
+        self.get_raw_button(8)
+    }
+
+    pub fn get_axis_button(&mut self, side: JoystickSide) -> bool {
+        match side {
+            JoystickSide::LeftHand => self.get_raw_button(9),
+            JoystickSide::RightHand => self.get_raw_button(10),
+        }
+    }
+
+    //axes
+    pub fn get_x(&mut self, side: JoystickSide) -> f32 {
+        match side {
+            JoystickSide::LeftHand => self.get_raw_axis(0),
+            JoystickSide::RightHand => self.get_raw_axis(4),
+        }
+    }
+
+    pub fn get_y(&mut self, side: JoystickSide) -> f32 {
+        match side {
+            JoystickSide::LeftHand => self.get_raw_axis(1),
+            JoystickSide::RightHand => self.get_raw_axis(5),
+        }
+    }
+
+    pub fn get_trigger(&mut self, side: JoystickSide) -> f32 {
+        match side {
+            JoystickSide::LeftHand => self.get_raw_axis(2),
+            JoystickSide::RightHand => self.get_raw_axis(3),
+        }
+    }
+
+    pub fn get_dpad(&mut self) -> DPAD {
+        match self.get_pov(1) {
+            0 => DPAD::Up,
+            1 => DPAD::UpRight,
+            2 => DPAD::Right,
+            3 => DPAD::DownRight,
+            4 => DPAD::Down,
+            5 => DPAD::DownLeft,
+            6 => DPAD::Left,
+            7 => DPAD::UpLeft,
+            _ => DPAD::Neutral,
+        }
+    }
+ }
+
+impl JoystickBase for XBoxController {
+    fn get_raw_axis(&mut self, axis: usize) -> f32 {
+        match self.ds.get_joystick_axis(self.port, axis) {
+            Ok(val) => val,
+            _ => 0f32,
```

Comment:
I don't necessarily agree with the premise that an error should result in a "neutral" value. I'd honestly prefer just returning a `Result` and having the user use `unwrap_or` to get a "neutral" value. It might make for a slightly more clunky API, but IMO the extra control that it gives the user is worth it. I'm open to discussion on the idea, but I'd like the error handling to remain consistent throughout the library. If we merged this now, there would be some APIs that just ignore errors (this) and others that return `Err`, which is inconsistent and something I'd rather avoid.